torch_geometric/nn/models/lightgcn.py

```diff
@@ -0,0 +1,301 @@
+
+        rankings = torch.mm(source_embeddings, target_embeddings.t())
+        top_indices = torch.topk(rankings, topK, dim=-1).indices
+        index_matrix = target_indices.reshape((1, -1))
+        index_matrix = index_matrix.repeat(source_indices.size(0), 1)
+
+        return index_matrix.gather(-1, top_indices)
+
+    def loss(self, preds: Tensor = None, edge_label: Tensor = None,
+             pos_ranks: Tensor = None, neg_ranks: Tensor = None
+             ) -> Tensor:
+        """Computes model loss based on objective.
+
+        Computes Bayesian Personalized Ranking (BPR) loss for ranking objective
+        and Binary Cross-Entropy (BCE) loss for link prediction objective.
+
+        .. note::
+
+            If objective is ranking, the i-th entry in the :obj:`pos_ranks`
+            vector and i-th entry in the :obj:`neg_ranks` entry must correspond
+            to positive and negative edges of the same entity (i.e. user).
+
+        Args:
+            preds (Tensor, optional): Predictions for link prediction objective
+                (default: :obj:`None`).
+            edge_label (Tensor, optional): Edge label for link prediction
+                objective (default: :obj:`None`).
+            pos_ranks (Tensor, optional): Positive edge rankings for ranking
+                objective (default: :obj:`None`).
+            neg_ranks (Tensor, optional): Negative edge rankings for ranking
+                objective (default: :obj:`None`).
+        """
+        if self.objective == LightGCN.objectives[0]:
+            assert (pos_ranks is not None and neg_ranks is not None), \
+                'For ranking objective, pos_ranks and neg_ranks must ' \
+                'be specified.'
+
+            return self.loss_fn(pos_ranks, neg_ranks, self.embeddings.weight)
+
+        else:
+            assert (preds is not None and edge_label is not None), \
+                'For link predictions objective, preds and edge_label must ' \
+                'be specified.'
+
+            return self.loss_fn(preds, edge_label.to(torch.float32))
+
+    def get_embeddings(self, edge_index: Adj,
+                       indices: LongTensor = None) -> Tensor:
+        """"""
+        if indices is None:
+            return self._propagate_embeddings(edge_index)
+
+        return self._propagate_embeddings(edge_index)[indices]
+
+    def _propagate_embeddings(self, edge_index: Adj) -> Tensor:
+        x = self.embeddings.weight
+        out = x * self.alpha[0]
+
+        for i in range(self.num_layers):
+            x = self.convs[i](x, edge_index)
+            out = torch.add(out, x * self.alpha[i + 1])
+
+        return out
+
+    def __repr__(self):
+        return (f'{self.__class__.__name__}('
+                f'num_layers={self.num_layers}, '
+                f'num_nodes={self.num_nodes}, '
+                f'embedding_dim={self.embedding_dim}, '
+                f'objective=\"{self.objective}\")')
+
+
+class BPRLoss(_Loss):
```

Comment:
Would it be easier to move this as a bpr_loss method inside the model class?

Reply:
Agree it might be easier but the advantage of a separate class it that it can inherit from PyTorch‘s general _Loss class and could in the future be replaced by a proper PyTorch implementation.